torch/distributed/_sharded_tensor/api.py

```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+                raise ValueError(
+                    f"Local shards' tensor dtype property need to be the same! "
+                    f"Found one local shard tensor dtype: {local_shards_dtype}, "
+                    f"the other local shard tensor dtype: {local_shard_tensor.dtype}"
+                )
 
-            if shard_meta.shard_lengths != list(local_shard_tensor.size()):
+            if local_shard_tensor.requires_grad != local_shards_requires_grad:
                 raise ValueError(
-                    f'Local shard tensor is incompatible with local ShardMetadata! '
-                    f'local shard tensor size: {local_shard_tensor.size()}, '
-                    f'local ShardMetadata shard lengths: {shard_meta.shard_lengths}'
+                    f"Local shards' tensor requires_grad property need to be the same! "
+                    f"Found one local shard tensor requires_grad: {local_shards_requires_grad}, "
+                    f"the other local shard tensor requires_grad: {local_shard_tensor.requires_grad}"
                 )
 
-            if local_shard_tensor.is_pinned() != tensor_properties.pin_memory:
+            if local_shard_tensor.is_pinned() != local_shards_is_pinned:
                 raise ValueError(
-                    f'Local shard tensor pin_memory does not match with tensor_properties! '
-                    f'local shard tensor pin_memory: {local_shard_tensor.is_pinned()}, '
-                    f'tensor_properties pin_memory: {tensor_properties.pin_memory}'
+                    f"Local shards' tensor pin_memory property need to be the same! "
+                    f"Found one local shard tensor is_pinned(): {local_shards_is_pinned}, "
+                    f"the other local shard tensor is_pinned(): {local_shard_tensor.is_pinned()}"
                 )
 
-            if local_shard_tensor.device != local_device:
+        # build a "local" ShardedTensorMetadata with all local shards on this rank
+        global_size = torch.Size(check_tensor_size(overall_size))
+
+        local_shard_tensor_metadata = ShardedTensorMetadata(
+            shards_metadata=local_shard_metadatas,
+            size=global_size,
+            tensor_properties=TensorProperties(
+                dtype=local_shards_dtype,
+                layout=local_shards_layout,
+                requires_grad=local_shards_requires_grad,
+                memory_format=torch.contiguous_format,
+                pin_memory=local_shards_is_pinned
+            )
+        )
+
+        # do all_gather to gather all sharded tensor metadatas in each rank
+        all_metadatas: List[Optional[ShardedTensorMetadata]] = [None for _ in range(world_size)]
+
+        dist.all_gather_object(
+            all_metadatas,
+            local_shard_tensor_metadata,
+            group=sharded_tensor._process_group
+        )
+
+        # do validation with gathered metadatas and build up a global sharded tensor metadata
```

Comment:
3. Validate the gathered local_sharded_metadata and build up a global sharded_tensor_metadata